### PR TITLE
Fix `Branch.bypass_pull_request_allowances` failing with "nil is not an object"

### LINKED ISSUE #2578: Calling `Branch.edit_protection` fails when any `required_pull_request_reviews` field set
<strike>It seems that the signature for `Branch.edit_protection` differs in the `.pyi` and `.py` file. In the [`.pyi`](https://github.com/PyGithub/PyGithub/blob/094538e10c1da6f3bbd31ea1f2d102e48e3649c5/github/Branch.pyi#L38) file the arguments are called `user_bypass_pull_request_allowances`, `team_bypass_pull_request_allowances` and `app_bypass_pull_request_allowances` (not the leading singular) whereas in the [.py`](https://github.com/PyGithub/PyGithub/blob/094538e10c1da6f3bbd31ea1f2d102e48e3649c5/github/Branch.py#L131C4-L131C4) file the arguments are plural (`users_bypass_pull_request_allowances`, `teams_bypass_pull_request_allowances` and `apps_bypass_pull_request_allowances`)</strike> fixed in #2571.

There also seems to be a problem where passing nothing for `user_bypass_pull_request_allowances` results in an error when `require_code_owner_reviews` or `dismiss_stale_reviews` are given:

```
github.GithubException.GithubException: 422 {"message": "Invalid request.\n\nNo subschema in \"anyOf\" matched.\nFor 'properties/bypass_pull_request_allowances', nil is not an object.\nNot all subschemas of \"allOf\" matched.\nFor 'anyOf/1', {\"dismiss_stale_reviews\"=>true, \"require_code_owner_reviews\"=>true, \"bypass_pull_request_allowances\"=>nil} is not a null.", "documentation_url": "https://docs.github.com/rest/branches/branch-protection#update-branch-protection"}
```

https://docs.github.com/en/rest/branches/branch-protection?apiVersion=2022-11-28#update-branch-protection

### PR DESCRIPTION
**Pull request by ThomasDevoogdt**
message:

    Invalid request.

    No subschema in "anyOf" matched.
    For 'properties/bypass_pull_request_allowances', nil is not an object.
    Not all subschemas of "allOf" matched.
    For 'anyOf/1', {"required_approving_review_count"=>1, "bypass_pull_request_allowances"=>nil} is not a null.

documentation_url:

https://docs.github.com/rest/branches/branch-protection#update-branch-protection

Fix by not adding the bypass_pull_request_allowances object at all if not needed. This is similar to the dismissal_restrictions logic.

Fixes #2578.

### CODE REVIEW

File: github/Branch.py [hunk 1]
```diff
@@ -298,23 +298,19 @@ def edit_protection(
...
                     "bypass_pull_request_allowances"
                 ] = {}
-                if users_bypass_pull_request_allowances is github.GithubObject.NotSet:
-                    users_bypass_pull_request_allowances = []
-                if teams_bypass_pull_request_allowances is github.GithubObject.NotSet:
-                    teams_bypass_pull_request_allowances = []
-                if apps_bypass_pull_request_allowances is github.GithubObject.NotSet:
-                    apps_bypass_pull_request_allowances = []
+
+            if users_bypass_pull_request_allowances is not github.GithubObject.NotSet:
```
Comment (trim21) [inline github/Branch.py:302]: Can we create a variable `bypass_pull_request_allowances` and split these in to 3 `if`?

from 
```python
if ... or ... or ... :
  if ...:
  if ...:
  if ...:
else:
   ...
```
to

```
bypass_pull_request_allowances = {}
if ...:
    ...
if ...:
    ... 
if ...:
    ...
if bypass_pull_request_allowances:
    ...
```

(I don't have write permission to this repo and I can't merge your PR, this is just my personal suggestion)

**Comment by EnricoMi** [comment]
But the documentation says `bypass_pull_request_allowances` is required. The `dismissal_restrictions` says, use an empty object, but you are not having any `dismissal_restrictions` or `bypass_pull_request_allowances` field.

Are the test replay data recorded or handcrafted?

**Comment by ThomasDevoogdt** [comment]
> But the documentation says `bypass_pull_request_allowances` is required. The `dismissal_restrictions` says, use an empty object, but you are not having any `dismissal_restrictions` or `bypass_pull_request_allowances` field.
> 
> Are the test replay data recorded or handcrafted?

I've adapted the replay data by hand in order to get the CI test passed. I run this code for a while in production and seems to do the job.

File: github/Branch.py (hunk 1 above)
Comment (EnricoMi) [reply github/Branch.py:302]: I like that. @ThomasDevoogdt would you be happy to rework these ifs that construct elements `required_pull_request_reviews`, `dismissal_restrictions` and `bypass_pull_request_allowances`, as suggested?

Comment (ThomasDevoogdt) [reply github/Branch.py:284]: Reworked the commit that relates to `bypass_pull_request_allowances`, and added an extra commit that relates to `dismissal_restrictions`.

**Review by trim21 — APPROVED** [review]
lgtm

**Review by EnricoMi — APPROVED** [review]
LGTM!

**Comment by jsnod** [comment]
@EnricoMi Is it possible to get this backported to `v1.59`? We attempted upgrading to `v1.59.1` for some new features, but it is unusable with this issue, and `v2.0` (having this fix) isn't released yet.

**Pull request merged**